development/extensions/tutorial_key_concepts.rst

```diff
@@ -0,0 +1,316 @@
+======================
+Tutorial: Key Concepts
+======================
+
+Introduction
+============
+
+Before we can jump into making our extension do great things, we need to cover some key concepts and
+practices used in phpBB.
+
+This tutorial explains:
+
+ * `Dependency injection`_
+ * `PHP files`_
+ * `Template files`_
+ * `Language files`_
+ * `Javascript and CSS files`_
+
+
+Dependency injection
+====================
+
+A large portion of the phpBB codebase since 3.1 is based on a fundamental mechanism, namely dependency injection. The
+goal is to not have to worry about managing the different dependencies of the components you want to use (e.g.: the
+database layer, the cache system or the current user object). It's also a means to cease using globally scoped
+variables.
+
+Since phpBB 3.1 this has been implemented using the Symfony DependencyInjection Component. All services provided by
+phpBB are defined in a YAML format file in ``/config/services.yml``. Extensions can use the services defined by phpBB as
+well as define their own services for their PHP classes in their own ``config/services.yml``.
+
+To provide a simple example of dependency injection in action, let's look at the constructor method for Acme Demo's
+controller class ``main.php``:
+
+.. code-block:: php
+
+    public function __construct(\phpbb\config\config $config, \phpbb\controller\helper $helper, \phpbb\template\template $template, \phpbb\user $user)
+    {
+        $this->config = $config;
+        $this->helper = $helper;
+        $this->template = $template;
+        $this->user = $user;
+    }
+
+As you can see, it requires the ``$config``, ``$helper``, ``$template``, and ``$user`` objects
+(aka services). To have these services automatically resolved by dependency injection, The Acme Demo extension must
+define its controller class as a service in the following manner:
+
+.. code-block:: yaml
+
+    services:
+        acme.demo.controller:
+            class: acme\demo\controller\main
+            arguments:
+                - '@config'
+                - '@controller.helper'
+                - '@template'
+                - '@user'
+
+Note that ``acme.demo.controller`` is the unique service name for the class. Service names
+can be whatever you want, but phpBB extensions should prefix them with the vendor and extension name (in this case,
```

Comment:
I would remove the can be whatever what you want line, as this is not the case for phpBB.